Generalize part of build process to actions (#42)
SneaksAndData/github-actions: README.md

10. [deploy_data_schemas_to_azfs](#deploy_data_schemas_to_azfs)
11. [run_azcopy](#run_azcopy)
12. [get_azure_share_sas](#get_azure_share_sas)

> 13. [setup_app](#setup_app)

Reviewer: ```suggestion
13. [setup_gh_app](#setup_gh_app)
```

14. [update_airflow_variables](#update_airflow_variables)
15. [create_pr](#create_pr)

> 16. [deploy_with_github_workflow](#deploy_with_github_workflow)

Reviewer: github -> gh

| project_version   | Version of the project                                                                      | False    |               |
| working_directory | Directory with airflow-variables GitHub repo                                                | False    |               |
| access_token      | An access token with push and create pull request permissions                               | False    |               |
| merge             | True if action should merge changes. Otherwise just create a branch and open a Pull Request | False    |               |

### Outputs
No outputs defined

### Usage
```yaml
name: Create pull request

on:
  workflow_dispatch:

jobs:
  contribute_changes:
    runs-on: ubuntu-latest
    steps:
      - name: Contribute Variable Changes
        uses: SneaksAndData/github-actions/contribute_changes@v0.0.17
        with:
          project_name: dbt-project
          working_directory: github-repository
          project_version: 0.0.1
          access_token: ${{ secrets.ACCESS_TOKEN }}
          merge: false
```

## activate_workflow

> Create pull request in repository

Reviewer: ```suggestion
Triggers a specified GitHub Workflow file with parameters.
```

### Inputs
| Name                           | Description                                                   | Optional | Default Value               |
|--------------------------------|:--------------------------------------------------------------|----------|-----------------------------|
| project_name                   | Name of the project                                           | False    |                             |
| project_version                | Version of the project                                        | False    |                             |
| working_directory              | Directory with airflow-variables GitHub repo                  | False    |                             |
| access_token                   | An access token with push and create pull request permissions | False    |                             |
| target_repo_name               | Repository to deploy                                          | False    |                             |
| deploy_environment             | Environment name to deploy                                    | True     | production                  |

> | generate_package_workflow_name | Name of workflow that generates deployment package            | True     | Prepare Helm Chart          |

Reviewer: these two are still relevant? these should be two different things

        with:
          access_token: ${{ secrets.ACCESS_TOKEN }}
          run_title: "Updating Project github-repo to version 1.1.1"
          repo_name: github-repo
          workflow_name: Prepare Helm chart
```

## read_airflow_variable

Read airflow variable, escape newlines for using content in other steps.

### Inputs
| Name               | Description                                       | Optional | Default Value |
|--------------------|:--------------------------------------------------|----------|---------------|
| project_name       | Project name                                      | False    |               |
| root_directory     | Root directory with variables repository          | False    |               |
| variables_sub_path | Subdirectory with JSON-encoded file with variable | False    |               |

### Outputs
| Name             | Description                   |
|------------------|:------------------------------|
| airflow_variable | JSON-encoded airflow variable |

### Usage
```yaml
name: Deploy latest tag

on:
  workflow_dispatch:

jobs:
  create_release:
    runs-on: ubuntu-latest
    steps:
      - name: Deploy variables

>         uses: SneaksAndData/github-actions/activate_workflow@v0.0.17

Reviewer: ```suggestion
        uses: SneaksAndData/github-actions/read_airflow_variable@v0.0.17
```

        with:
          access_token: ${{ secrets.ACCESS_TOKEN }}
          run_title: "Updating Project github-repo to version 1.1.1"
          repo_name: github-repo

>           workflow_name: Prepare Helm chart

Reviewer: looks like a copypaste from the previous one

          source_directory: source/directory/on/build/agent
          target: ${{ steps.sas.outputs.authorized_destination }}
```

## setup_app

### Description

> Configure git to use GitHub app in next steps of workflow. This strep is required for interation with GitHub api in the

Reviewer: ```suggestion
Configure git client in the workflow job to authenticate to GitHub using a GitHub App instead of builtin repo-scoped GITHUB_TOKEN.
```

| git_user_email      | User email for git client   | False    |               |
| git_user_name       | User name for git client    | False    |               |

### Outputs
| Name         | Description                          |
|--------------|--------------------------------------|
| access_token | Access token generated by GitHub API |

### Usage
```yaml
name: Checkout repo using github app

on:
  workflow_dispatch:

jobs:
  create_release:
    runs-on: ubuntu-latest
    steps:
      - name: Reconfigure Git to use App
        id: delamain_auth
        uses: SneaksAndData/github-actions/setup_app@v0.0.12
        with:
          app_private_key: ${{ env.APP_PRIVATE_KEY }}
          app_installation_id: ${{ env.APP_INSTALLATION_ID }}
          app_id: ${{ env.APP_ID }}
          git_user_name: "Github App"
          git_user_email: "user@example.com"

      - name: Checkout Airflow Variables
        uses: actions/checkout@v3
        with:
          repository: 'SneaksAndData/airflow-variables'
          token: ${{ steps.delamain_auth.outputs.access_token }}
          path: 'airflow-variables'
          fetch-depth: 0
```

## update_airflow_variables

> Update airflow variables to use selected project and graph, previously generated in workflow steps.

Reviewer: ```suggestion
Update airflow configuration of a specified dbt project to the current version (git tag) and provided execution graph
```

### Inputs
| Name              | Description                         | Optional | Default Value |
|-------------------|:------------------------------------|----------|---------------|
| project_name      | Name of the project                 | False    |               |
| project_version   | Version of the project              | False    |               |
| project_graph     | Graph generated by metadata-manager | False    |               |
| working_directory | User name for git client            | False    |               |

### Outputs
No outputs defined

### Usage
```yaml
name: Update airflow-variables

on:
  workflow_dispatch:

jobs:
  create_release:
    runs-on: ubuntu-latest
    steps:
      - name: Update Project Version
        uses: SneaksAndData/github-actions/update_airflow_variables@v0.0.12
        with:
          project_name: ${{ env.PROJECT_NAME }}
          working_directory: github-repository
          project_version: ${{steps.version.outputs.version}}
          project_graph: ${{steps.airflow_dag.outputs.PROJECT_GRAPH}}
```

## create_pr

Create pull request in repository

**NOTES**:

> 1) If deploy environment is not equal to `test`, this action creates pull request and merge it. Otherwise,

Reviewer: ```suggestion
1) If deploy environment is not equal to `test`, this action creates pull request and merges it. Otherwise,
```

it only pushes a new branch.

### Inputs
| Name               | Description                                                   | Optional | Default Value |
|--------------------|:--------------------------------------------------------------|----------|---------------|
| project_name       | Name of the project                                           | False    |               |
| project_version    | Version of the project                                        | False    |               |
| working_directory  | Directory with airflow-variables GitHub repo                  | False    |               |
| access_token       | An access token with push and create pull request permissions | False    |               |
| deploy_environment | Environment name to deploy                                    | False    |               |

### Outputs
No outputs defined

### Usage
```yaml
name: Create pull request

on:
  workflow_dispatch:

jobs:
  create_release:
    runs-on: ubuntu-latest
    steps:
      - name: Contribute Variable Changes
        uses: SneaksAndData/github-actions/create_pr@v0.0.12
        with:

>           project_name: ${{ env.PROJECT_NAME }}

Reviewer: put some dummy name here

          working_directory: github-repository

>           project_version: ${{steps.version.outputs.version}}

Reviewer: same here re generics

>           access_token: ${{ steps.delamain_auth.outputs.access_token }}

Reviewer: i think this should be more generic in the example

| project_name       | Name of the project                                           | False    |               |
| project_version    | Version of the project                                        | False    |               |
| working_directory  | Directory with airflow-variables GitHub repo                  | False    |               |
| access_token       | An access token with push and create pull request permissions | False    |               |
| target_repo_name   | Repository to deploy                                          | False    |               |
| deploy_environment | Environment name to deploy                                    | True     | production    |

### Outputs
No outputs defined

### Usage
```yaml
name: Deploy latest tag

on:
  workflow_dispatch:

jobs:
  create_release:
    runs-on: ubuntu-latest
    steps:
      - name: Deploy variables
        if: ${{ startsWith(github.ref, 'refs/tags') }}
        uses: SneaksAndData/github-actions/deploy_with_github_workflow@v0.0.12
        with:
          project_name: ${{ env.PROJECT_NAME }}
          working_directory: github-repository
          project_version: ${{steps.version.outputs.version}}

>           access_token: ${{ steps.delamain_auth.outputs.access_token }}

Reviewer: replace variables with hardcoded examples for brevity

          path: airflow-variables
          fetch-depth: 0
```

## update_airflow_variables

Update airflow configuration of a specified dbt project to the current version (git tag) and provided execution graph

### Inputs
| Name              | Description                         | Optional | Default Value |
|-------------------|:------------------------------------|----------|---------------|
| project_name      | Name of the project                 | False    |               |
| project_version   | Version of the project              | False    |               |
| project_graph     | Graph generated by metadata-manager | False    |               |
| working_directory | User name for git client            | False    |               |

### Outputs
No outputs defined

### Usage
```yaml
name: Update airflow-variables

on:
  workflow_dispatch:

jobs:
  create_release:
    runs-on: ubuntu-latest
    steps:
      - name: Update Project Version
        uses: SneaksAndData/github-actions/update_airflow_variables@v0.0.12
        with:
          project_name: dbt-project
          working_directory: github-repository
          project_version: 0.0.1
          project_graph: "{}"
```

> ## create_pr

Reviewer: too generic name for this action

|--------------------|:--------------------------------------------------------------|----------|---------------|
| project_name       | Name of the project                                           | False    |               |
| project_version    | Version of the project                                        | False    |               |
| working_directory  | Directory with airflow-variables GitHub repo                  | False    |               |
| access_token       | An access token with push and create pull request permissions | False    |               |
| deploy_environment | Environment name to deploy                                    | False    |               |

### Outputs
No outputs defined

### Usage
```yaml
name: Create pull request

on:
  workflow_dispatch:

jobs:
  create_release:
    runs-on: ubuntu-latest
    steps:
      - name: Contribute Variable Changes
        uses: SneaksAndData/github-actions/create_pr@v0.0.12
        with:
          project_name: dbt-project
          working_directory: github-repository
          project_version: 0.0.1
          access_token: ${{ secrets.ACCESS_TOKEN }}
          deploy_environment: production
```

> ## deploy_with_gh_workflow

Reviewer: incorrect name. action should be just activating GH workflow in a repo

Create pull request in repository

**NOTES**:
1) This action requires workflows with names `Prepare Helm Chart` and `Deploy vairables to Airflow` to be

> present in repository

Reviewer: untrue. you should be able to use this action to activate any workflow

### Inputs
| Name               | Description                                                   | Optional | Default Value |
|--------------------|:--------------------------------------------------------------|----------|---------------|
| project_name       | Name of the project                                           | False    |               |
| project_version    | Version of the project                                        | False    |               |
| working_directory  | Directory with airflow-variables GitHub repo                  | False    |               |
| access_token       | An access token with push and create pull request permissions | False    |               |
| target_repo_name   | Repository to deploy                                          | False    |               |

> | deploy_environment | Environment name to deploy                                    | True     | production    |

Reviewer: missing workflow name here

### Outputs
No outputs defined

### Usage
```yaml
name: Deploy latest tag

on:
  workflow_dispatch:

jobs:
  create_release:
    runs-on: ubuntu-latest
    steps:
      - name: Deploy variables
        if: ${{ startsWith(github.ref, 'refs/tags') }}
        uses: SneaksAndData/github-actions/deploy_with_gh_workflow@v0.0.12
        with:
          project_name: dbt-project
          working_directory: github-repository
          project_version: 0.0.1
          access_token: ${{ secrets.ACCESS_TOKEN }}
```

> ## export_dag_and_sql

Reviewer: this is dbt-specific, should include dbt in the name

Export airflow dag and sql files for use in [update_airflow_variables](#update_airflow_variables) step. 

### Inputs
| Name               | Description                                                   | Optional | Default Value |
|--------------------|:--------------------------------------------------------------|----------|---------------|
| deploy_environment | Environment name to deploy                                    | True     | production    |
| project_name       | Name of the project                                           | False    |               |

### Outputs
| Name        | Description                    |
|-------------|--------------------------------|
| airflow_dag | AirFlow dag serialized as JSON |

### Usage
````yaml
name: Generate Airflow DAG

on:
  workflow_dispatch:

jobs:
  create_release:
    runs-on: ubuntu-latest
    steps:
      - name: Export airflow dag and sql files
        uses: SneaksAndData/github-actions/export_dag_and_sql@v0.0.12
        with:
          deploy_environment: test
          project_name: dbt_project
        id: airflow_dag
````

> ## run_dbt_compile

Reviewer: i don't see why this should be an action?
